### PR TITLE
Add support for jira or other issue trackers.

### PR DESCRIPTION
**Pull request by gseguin**

### CODE REVIEW

**Comment by gseguin** [comment]
The failure in travis is caused by node 0.8 other versions of node are passing.

**Comment by scottgonzalez** [comment]
Can you create a `Changelog.ticketParsers` object that's keyed off of the ticket type (github, jira) and the value is a function that you're currently assigning to `getTicketRefs()`? We'll need to change the `ticketUrl` option to accept a hash that's also keyed off of the ticket type. We'll also keep the string version for simplicity; in the case that a string is provided, we'll assume that it should be used for every match, regardless of ticket type (except for the external GitHub references which are hard-coded to GitHub URLs already).

The goal of this change is to allow a project to have multiple ticketing systems (I know it's rare, but it does happen). In order to make this part work, we'll need to add another option (perhaps `ticketTypes`) that defines which parsers should be used (as an array). We'll then run each parser listed in this option against each commit, in the order specified. If we default this option to `["github", "jira"]` then this should work for both types out of the box (assuming you're only using one of those, you can just set `ticketUrl` to a string as you're doing now).

Thoughts?

**Comment by gseguin** [comment]
Thanks Scott.
Will give it a shot, test it with my current setup and update the PR.

**Comment by scottgonzalez** [comment]
Thanks. Let me know if you have any questions or want to pair on this.

File: index.js [hunk 1]
```diff
@@ -35,8 +116,18 @@ Changelog.prototype.parse = function( callback ) {
 	}.bind( this ));
 };
 
-Changelog.prototype.ticketUrl = function( id ) {
-	return this.options.ticketUrl.replace( "{id}", id );
+Changelog.prototype.ticketUrl = function( id, ticketType ) {
+	if ( !ticketType ) {
```
Comment (gseguin) [inline index.js]: Wonder if I should add a check for ```this.ticketTypes.length === 1```, that would be safer

File: index.js [hunk 2]
```diff
@@ -17,6 +17,38 @@ function Changelog( options ) {
...
 
+	var ticketTypes;
+	if ( options.ticketTypes ) {
+		ticketTypes = options.ticketTypes;
+	} else {
+		ticketTypes = [ "github" ];
+	}
+
+	this.ticketTypes = [];
+	// Re-order ticketTypes according to the key order in Changelog.ticketParsers
```
Comment (scottgonzalez) [inline index.js]: `ticketTypes` needs to be an option. It is what defines which parsers get used and what order to use them in.

File: index.js [hunk 3]
```diff
@@ -17,6 +17,38 @@ function Changelog( options ) {
 	this.options = options;
 	this.repo = new Repo( this.options.repo );
 
+	var ticketTypes;
+	if ( options.ticketTypes ) {
+		ticketTypes = options.ticketTypes;
+	} else {
+		ticketTypes = [ "github" ];
```
Comment (scottgonzalez) [inline index.js]: The default should be `[ "github", "jira" ]`.

Make an object of default options and merge it into the provided options.

File: index.js [hunk 4]
```diff
@@ -17,6 +17,38 @@ function Changelog( options ) {
...
+	// Re-order ticketTypes according to the key order in Changelog.ticketParsers
+	Object.keys( Changelog.ticketParsers).forEach(
+		function( key ) {
+			if ( ticketTypes.indexOf( key ) != -1 ) {
+				this.ticketTypes.push( key );
+			}
+		}.bind(this)
+	);
+
+	if ( "string" === typeof options.ticketUrl ) {
```
Comment (scottgonzalez) [inline index.js]: No yoda conditionals. Reverse the values.

File: index.js [hunk 5]
```diff
@@ -17,6 +17,38 @@ function Changelog( options ) {
...
+		// backward compatibility
```
Comment (scottgonzalez) [inline index.js]: This isn't back-compat, it's a nicer API for single values.

File: index.js [hunk 6]
```diff
@@ -17,6 +17,38 @@ function Changelog( options ) {
...
+		if ( this.ticketTypes.length > 1 ) {
+			throw new Error(
+				this.ticketTypes.length +
+				" ticket types were sent but only one ticket url template was. " +
+				"[ticketUrl should be an Object]"
+			);
+		}
```
Comment (scottgonzalez) [inline index.js]: I don't think we need any checks like this.

File: index.js [hunk 7]
```diff
@@ -25,6 +57,55 @@ function Changelog( options ) {
 	}
 }
 
+Changelog.ticketParsers = {
+	// Order of keys is important!
+	// When passed several ticketTypes, they will be re-ordered in the order of the keys of this map
+	// This works around github's (or other) regex to be greedy
```
Comment (scottgonzalez) [inline index.js]: Order is irrelevant here. I'm also not sure what's greedy about the GitHub regex; it should never match a JIRA ticket.

File: index.js (hunk 1 above)
Comment (scottgonzalez) [inline index.js]: I'm not sure why you'd want to check that.

File: index.js [hunk 8]
```diff
@@ -35,8 +116,18 @@ Changelog.prototype.parse = function( callback ) {
...
-	return this.options.ticketUrl.replace( "{id}", id );
+Changelog.prototype.ticketUrl = function( id, ticketType ) {
+	if ( !ticketType ) {
+		ticketType = this.ticketTypes[0];
+	}
+
+	return this.ticketUrlTemplates[ ticketType ].replace( "{id}", id );
+};
+
+Changelog.ticketUrlFactory = function( ticketType, thisArg ) {
```
Comment (scottgonzalez) [inline index.js]: I don't think we need factories.

File: package.json [hunk 9]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "changelogplease",
-  "version": "1.1.1",
+  "version": "1.2.0",
```
Comment (scottgonzalez) [inline package.json]: Please don't change versions. That's the maintainer's job and is often scripted as part of a release process.

File: index.js (hunk 2 above)
Comment (gseguin) [inline index.js]: It is an option: https://github.com/scottgonzalez/changelogplease/pull/8/files#diff-168726dbe96b3ce427e7fedce31bb0bcR21

File: index.js (hunk 7 above)
Comment (gseguin) [inline index.js]: The tests revealed github regex matching a JIRA ticket hence so either something is wrong with the regex or it is indeed greedy.

Disable the reordering and run the tests to observe.

File: index.js (hunk 8 above)
Comment (gseguin) [inline index.js]: So you want to create the wrapping function 'manually' in the forEach loop: https://github.com/scottgonzalez/changelogplease/pull/8/files#diff-168726dbe96b3ce427e7fedce31bb0bcR188 ?

**Comment by gseguin** [comment]
bump

File: index.js (hunk 2 above)
Comment (scottgonzalez) [inline index.js]: But why are you regenerating the list with a potentially different order? The option is what defines the order.

File: index.js (hunk 8 above)
Comment (scottgonzalez) [inline index.js]: I don't know why you need a wrapping function in the first place.You're generating a function that has one line of code in it.

**Comment by gseguin** [comment]
Scott, do these last changes look good to you?

**Comment by scottgonzalez** [comment]
Ugh, how did I let a whole month go by? I'll try to get back to this today.

**Comment by gseguin** [comment]
No worries, in the meantime I'm using my fork.

File: index.js (hunk 7 above)
Comment (scottgonzalez) [inline index.js]: I switched this so that github came first and all the tests passed.

File: index.js (hunk 1 above)
Comment (scottgonzalez) [inline index.js]: I'm still not sure why you think this should be handled.

File: index.js [hunk 10]
```diff
@@ -25,6 +38,55 @@ function Changelog( options ) {
...
+Changelog.ticketParsers = {
+	// Order of keys is important!
+	// When passed several ticketTypes, they will be re-ordered in the order of the keys of this map
+	// This works around github's (or other) regex to be greedy
+	jira: function ( commit, ticketUrl ) {
+		var tickets = [];
+
+		// Sane global exec with iteration over matches
+		commit.replace(
+			/Fix(?:e[sd])? ((?:([A-Z]{1,10})-?)[A-Z]+-\d+)/g,
```
Comment (scottgonzalez) [inline index.js]: How did you come up with this regex? It doesn't seem to comply with the JIRA project key format.

**Comment by scottgonzalez** [comment]
I created a PR against your branch with some changes: https://github.com/gseguin/changelogplease/pull/1

**Comment by gseguin** [comment]
Merged gseguin/changelogplease#1

**Comment by scottgonzalez** [comment]
I just wrote the docs for this and merged into master. Can you review the docs and make sure I didn't miss anything? I'll release a new version as soon as the docs are reviewed. Thanks!

**Comment by scottgonzalez** [comment]
Released in v1.2.0.

:star2: Thanks again! :star2: 

@arschmitz and I have ideas for more customizations which might be useful for you. See #9 and #10.

**Pull request closed without merging**